tests/subsys/audio_module/src/functional_test.c

```diff
@@ -0,0 +1,1801 @@
+		zassert_mem_equal(&mod_functions_null, &test_mod_functions,
+				  sizeof(struct audio_module_functions),
+				  "Failed close, modified the modules functions");
+		zassert_mem_equal(&mod_parameters, &test_mod_parameters,
+				  sizeof(struct audio_module_parameters),
+				  "Failed close, modified the modules parameter settings");
+		zassert_mem_equal(&mod_context, &test_mod_context, sizeof(struct mod_context),
+				  "Failed close, modified the modules context");
+	}
+}
+
+ZTEST(suite_audio_module_functional, test_open_fnct)
+{
+	int ret;
+	char *test_inst_name = "TEST instance 1";
+
+	mod_description.functions = &mod_functions_populated;
+
+	/* Fake internal empty data FIFO success */
+	data_fifo_init_fake.custom_fake = fake_data_fifo_init__succeeds;
+
+	data_fifo_deinit(&mod_fifo_rx);
+	data_fifo_deinit(&mod_fifo_tx);
+
+	data_fifo_init(&mod_fifo_rx);
+	data_fifo_init(&mod_fifo_tx);
+
+	for (int i = AUDIO_MODULE_TYPE_INPUT; i <= AUDIO_MODULE_TYPE_IN_OUT; i++) {
+		test_context_set(&test_mod_context, &mod_config);
+
+		mod_parameters.thread.msg_rx = &mod_fifo_rx;
+		mod_parameters.thread.msg_tx = &mod_fifo_tx;
+
+		mod_description.type = i;
+
+		memset(&handle, 0, sizeof(struct audio_module_handle));
+
+		handle.state = AUDIO_MODULE_STATE_UNDEFINED;
+
+		ret = audio_module_open(
+			&mod_parameters, (struct audio_module_configuration *)&mod_config,
+			test_inst_name, (struct audio_module_context *)&mod_context, &handle);
+
+		zassert_equal(ret, 0, "Open function did not return successfully (0): ret %d", ret);
+		zassert_equal(handle.state, AUDIO_MODULE_STATE_CONFIGURED,
+			      "Open state not AUDIO_MODULE_STATE_CONFIGURED (%d) rather %d",
+			      AUDIO_MODULE_STATE_CONFIGURED, handle.state);
+		zassert_mem_equal(&handle.name, test_inst_name, sizeof(test_inst_name),
+				  "Failed open, name should be %s, but is %s", test_inst_name,
+				  handle.name);
+		zassert_mem_equal(handle.description, &mod_description,
+				  sizeof(struct audio_module_description),
+				  "Failed open, module descriptions differ");
+		zassert_mem_equal(handle.description->functions, &mod_functions_populated,
+				  sizeof(struct audio_module_functions),
+				  "Failed open, module function pointers differ");
+		zassert_mem_equal(&handle.thread, &mod_parameters.thread,
+				  sizeof(struct audio_module_thread_configuration),
+				  "Failed open, module thread settings differ");
+		zassert_mem_equal(handle.context, &test_mod_context, sizeof(struct mod_context),
+				  "Failed open, module contexts differ");
+		zassert_equal(handle.use_tx_queue, false,
+			      "Open failed use_tx_queue is not false: %d", handle.use_tx_queue);
+		zassert_equal(handle.dest_count, 0, "Open failed dest_count is not 0: %d",
+			      handle.dest_count);
+
+		k_thread_abort(handle.thread_id);
+	}
+
+	for (int i = AUDIO_MODULE_TYPE_INPUT; i <= AUDIO_MODULE_TYPE_IN_OUT; i++) {
+		test_context_set(&test_mod_context, &mod_config);
+
+		mod_parameters.thread.msg_rx = NULL;
```

Comment:
Seems to me that the tests are repeated with msg_rx set to NULL. Is this needed?

Reply:
If it is, we can let msg_rx and tx take an array of NULL or mod_fifo_tx/mod_fifo_rx. That would remove a significant number of lines.